### PR TITLE
Add helix event sending, support docker args in simpleDockerNode, misc other utils

### PR DESCRIPTION
**Pull request by chcosta**
- Update Pipeline.groovy to support `getValidJobNameString` as a static method so that we can later derive the job name.  I made a separate change which made this irrelevant where I explicitly specify a job name rather than letting it derive.  The result is that being able to determine the job name via a static method is not necessary.  I left this change in my PR, but I'm ok with reverting it if it's a concern.
- Update Pipeline.groovy to allow you to explicitly specify job names when creating triggers
- Add Helix event logic so that we can send Helix events.  I added the Helix event logic here (rather than BuildTools) so that we can send events from anywhere during the pipeline
- Update simpleDockerNode so that we can pass arguments to the docker command

### CODE REVIEW

File: src/org/dotnet/ci/pipelines/Pipeline.groovy [hunk 1]
```diff
@@ -347,13 +347,17 @@ class Pipeline {
...
         // Job name is based off the parameters 
 
         def isPR = triggerBuilder.isPRTrigger()
-        def jobName = getPipelineJobName(params)
-        def fullJobName = Utilities.getFullJobName(jobName, isPR)
+        def _jobName = jobName
+        def fullJobName = jobName
+        if(_jobName == null) {
+            _jobName = Pipeline.getPipelineJobName(_baseJobName, params)
+            fullJobName = Utilities.getFullJobName(_jobName, isPR)
```
Comment (mmitche) [inline src/org/dotnet/ci/pipelines/Pipeline.groovy]: Any reason we aren't calling getFullJobName in the case of an explicit job name?  It is just the thing that adds _prtest in the case of pr.  Wouldn't that mean an overlap of job names between trigger methods?

File: vars/createSender.groovy [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+}
+def call(String correlationId, String credentialsIdentifier) {
+    return call(correlationId, null, credentialsIdentifier)
+}
+
+/* Create a helix event sender which is used to send events to the endpoint specified in the
+ * credentialsIdentifier
+ */
+def call(String correlationId, String workItemId, String credentialsIdentifier) {
+    withCredentials([string(credentialsId: credentialsIdentifier, variable: 'HelixEndpoint')]) {
```
Comment (mmitche) [inline vars/createSender.groovy]: The credentialsId here appears unused.  Is this intentional or is it used in some other odd way?

File: vars/getBuildNumber.groovy [hunk 3]
```diff
@@ -0,0 +1,7 @@
+/**
+  * Retrieves formatted official build number.
+  */
+def call() {
+    def versionNumber = VersionNumber('${BUILD_YEAR}${BUILD_MONTH, XX}${BUILD_DAY, XX}-${BUILDS_TODAY, XX}')
```
Comment (mmitche) [inline vars/getBuildNumber.groovy]: For debugging, can you put asserts in here to ensure that the environment variables you are depending on are installed (along with a note of the required plugin upon failed assertion)?

File: vars/simpleDockerNode.groovy [hunk 4]
```diff
@@ -25,6 +25,10 @@ def call(String dockerImageName, Closure body) {
 //  hostVersion - Host VM version.  See Agents.getDockerMachineAffinity for explanation.
 //  body - Closure, see example
 def call(String dockerImageName, String hostVersion, Closure body) {
+  call(dockerImageName, hostVersion, '', body)
```
Comment (mmitche) [inline vars/simpleDockerNode.groovy]: Nit: indent formatting

File: vars/sendEvent.groovy [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import org.dotnet.helix.*
+
+def call(String type, Map data, Map senderMap) {
+    def sender = createSender(senderMap)
+    def eventData = new HelixEventData()
+    eventData.Type = type
+    eventData.Data = data
+    return sender.Send(eventData)
```
Comment (mmitche) [inline vars/sendEvent.groovy]: Do we need error handling here? It looks like error handling is added inside here, but it rethrows the exception. Is this the desired behavior?

File: vars/simpleDockerNode.groovy (hunk 4 above)
Comment (mmitche) [inline vars/simpleDockerNode.groovy]: Nit: Indent formatting.

File: src/org/dotnet/helix/NoopSender.groovy [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package org.dotnet.helix;
+
+import java.util.UUID
+
+class NoopSender implements IHelixEventSender {
+    public final String CorrelationId
```
Comment (mmitche) [inline src/org/dotnet/helix/NoopSender.groovy]: Nit: typically in Java the member variables have their first letter lower case.

File: vars/createSender.groovy (hunk 2 above)
Comment (chcosta) [reply vars/createSender.groovy]: the credential is used on the next line 

`HelixEvent.Initialize(env.HelixEndpoint) `

Comment (mmitche) [reply vars/createSender.groovy]: Oh, I'm an idiot...i mistakenly read that as another credential binding

**Comment by mmitche** [comment]
@dotnet-bot test this please (@chcosta ignore)

**Comment by mmitche** [comment]
@dotnet-bot test this please (@chcosta ignore)

**Comment by chcosta** [comment]
@dotnet-bot test this please

**Comment by chcosta** [comment]
@mmitche , the "CI Tests" failure don't appear to be related to my changes.  They appear to be related to failures in the Helix runs.  I don't have much experience here though, so I'm happy to dig in if you think I should block on these failures.

**Comment by mmitche** [comment]
@chcosta I don't think that's related.

**Pull request merged**